tests/bsim/bluetooth/host/misc/sample_test/src/peer.c

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/att.h>
+#include <zephyr/bluetooth/gatt.h>
+
+#include "testlib/adv.h"
+#include "testlib/att_read.h"
+#include "testlib/att_write.h"
+#include "testlib/conn.h"
+
+#include "bs_helpers/flags.h"
+#include "bs_helpers/sync.h"
+#include "bs_helpers/testcase.h"
+#include "bs_helpers/timeout.h"
+
+/* local includes */
+#include "data.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(peer, CONFIG_TEST_APP_LOG_LEVEL);
+
+static DEFINE_FLAG(is_subscribed);
+static DEFINE_FLAG(got_notification_1);
+static DEFINE_FLAG(got_notification_2);
+
+int find_characteristic(struct bt_conn *conn,
+			const struct bt_uuid *svc,
+			const struct bt_uuid *chrc,
+			uint16_t *chrc_value_handle)
+{
+	uint16_t svc_handle;
+	uint16_t svc_end_handle;
+	uint16_t chrc_end_handle;
+	int err;
+
+	LOG_DBG("");
+
+	err = bt_testlib_gatt_discover_primary(&svc_handle, &svc_end_handle, conn, svc, 1, 0xffff);
+	if (err) {
+		LOG_ERR("Failed to discover service: %d", err);
+
+		return err;
+	}
+
+	LOG_DBG("svc_handle: %u, svc_end_handle: %u", svc_handle, svc_end_handle);
+
+	err = bt_testlib_gatt_discover_characteristic(chrc_value_handle, &chrc_end_handle,
+						      NULL, conn, chrc, (svc_handle + 1),
+						      svc_end_handle);
+	if (err) {
+		LOG_ERR("Failed to get value handle: %d", err);
+
+		return err;
+	}
+
+	LOG_DBG("chrc_value_handle: %u, chrc_end_handle: %u", *chrc_value_handle, chrc_end_handle);
+
+	return err;
+}
+
+static uint8_t received_notification(struct bt_conn *conn,
+				     struct bt_gatt_subscribe_params *params,
+				     const void *data,
+				     uint16_t length)
+{
+	if (length) {
+		LOG_INF("RX notification");
+		LOG_HEXDUMP_DBG(data, length, "payload");
+
+		TEST_ASSERT(length == sizeof(payload_1), "Unexpected length: %d\n", length);
```

Comment:
Should this check be moved to before the `if (length)` check? Or do we allow empty notifications in this test?

Reply:
There some weird quirk of the API where this gets called with length 0 when you unsubscribe, at least on disconnect.